macros/src/newtype.rs

```diff
@@ -0,0 +1,82 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Data, DeriveInput, Fields};
+
+pub fn cw_storage_newtype_key_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let expanded = impl_newtype(&input);
+
+    TokenStream::from(expanded)
+}
+
+fn impl_newtype(input: &DeriveInput) -> proc_macro2::TokenStream {
+    // Extract the struct name
+    let name = &input.ident;
+
+    // Extract the inner type
+    let inner_type = match &input.data {
+        Data::Struct(data_struct) => match &data_struct.fields {
+            Fields::Unnamed(fields) if fields.unnamed.len() == 1 => Some(&fields.unnamed[0].ty),
+            _ => None,
+        },
+        _ => None,
+    };
+
+    let inner_type = inner_type
+        .expect("NewTypeKey can only be derived for newtypes (tuple structs with one field)");
```

Comment:
There are ways to display macro errors similarly to other rust compiler errors, with source code being highlighted and all. I don't think it's a must, but if you're interested: https://docs.rs/syn/latest/syn/struct.Error.html

Reply:
I can probably find an example later

Reply:
[Here's how Sylvia does it](https://github.com/CosmWasm/sylvia/blob/10e8eb9ce722c4e2f2bed6bce950acbcefd2b937/sylvia-derive/src/lib.rs#L243-L258)

Reply:
thanks, I don't have a lot of experience writing macros, happy to learn better techniques. Reminding myself to address this and other comments early next week if not later today :)

Reply:
done in https://github.com/CosmWasm/cw-storage-plus/pull/88/commits/afa93b6840118cccfeae1fe550c20899cdb87f5c